### PR TITLE
Add additional auth support to Gitlab (#705)

### LINKED ISSUE #705: Gitlab modules does not support oauth
##### SUMMARY
GitLab API v4 doesn't support login/password support since GitLab 10.2.
However it still support oauth token that we can get from login/password

All GitLab modules should support login/password with oauth mechanism.

##### ISSUE TYPE
Feature Idea

##### COMPONENT NAME
gitlab_groups
gitlab_deploy_key
gitlab_hook
gitlab_project
gitlab_project_variable
gitlab_runner
gitlab_user

##### ADDITIONAL INFORMATION
python-gitlab API support oauth token.
We should only find a way to gather the token from login/password

https://github.com/ansible/ansible/issues/61525

### PR DESCRIPTION
**Pull request by josham**
##### SUMMARY
Add the following additional ways to authenticate:
- username/password via oauth_token
- oauth_token
- job_token
Fixes #705

Other small changes:
- removed unused imports from module_utils.gitlab
- fix bug in gitlab_project to check if avatar_path is provided

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME

gitlab_branch
gitlab_group_members
gitlab_group_variable
gitlab_project_members
gitlab_project_variable
gitlab_runner
gitlab_deploy_key
gitlab_group
gitlab_hook
gitlab_project
gitlab_protected_branch
gitlab_user

### CODE REVIEW

File: plugins/module_utils/gitlab.py [hunk 1]
```diff
@@ -7,23 +7,23 @@
...
 except ImportError:
-    from urllib.parse import quote_plus  # Python 3+
+    from urllib.parse import quote_plus, urljoin  # Python 3+
 
 import traceback
 
 GITLAB_IMP_ERR = None
 try:
     import gitlab
+    import requests
```
Comment (russoz) [inline plugins/module_utils/gitlab.py:26]: This add a dependency to the module, that might not even be required if the user is authenticating with `api_token`. 

At the very least, this new dependency should show up in the modules' documentation.

File: plugins/modules/source_control/gitlab/gitlab_deploy_key.py [hunk 2]
```diff
@@ -29,6 +29,14 @@
...
       - GitLab token for logging in.
     type: str
+  api_oauth_token:
+    description:
+      - GitLab OAuth token for logging in.
+    type: str
+  api_job_token:
+    description:
+      - GitLab CI job token for logging in.
+    type: str
```
Comment (russoz) [inline plugins/modules/source_control/gitlab/gitlab_deploy_key.py:39]: The Right Way(TM) to do that is creating a doc_fragment with this text and make the modules point to it. If not in this PR, then please raise an issue to address that.

**Review by russoz — COMMENTED** [review]
Hi @josham thanks for your contribution. The new dependency might need some rework, and maybe you could take the opportunity to deal with the doc_fragment - or not.

File: plugins/modules/source_control/gitlab/gitlab_branch.py [hunk 3]
```diff
@@ -132,12 +141,18 @@ def main():
         mutually_exclusive=[
             ['api_username', 'api_token'],
             ['api_password', 'api_token'],
+            ['api_username', 'api_oauth_token'],
```
[the remark this answers is not in the document]
Comment (josham) [reply plugins/modules/source_control/gitlab/gitlab_branch.py:144]: Will make this change

File: plugins/module_utils/gitlab.py (hunk 1 above)
Comment (josham) [reply plugins/module_utils/gitlab.py:26]: requests is a dependency of gitlab's python api so I thought it was OK to use it. Do you still want me to explicitly list it?

File: plugins/modules/source_control/gitlab/gitlab_deploy_key.py (hunk 2 above)
Comment (josham) [reply plugins/modules/source_control/gitlab/gitlab_deploy_key.py:39]: Will make this change

File: plugins/modules/source_control/gitlab/gitlab_project.py [hunk 4]
```diff
@@ -299,10 +310,11 @@ def create_or_update_project(self, project_name, namespace, options):
             project = self.create_project(namespace, project_options)
 
             # add avatar to project
-            try:
-                project.avatar = open(options['avatar_path'], 'rb')
-            except IOError as e:
-                self._module.fail_json(msg='Cannot open {0}: {1}'.format(options['avatar_path'], e))
+            if options['avatar_path']:
```
Comment (felixfontein) [inline plugins/modules/source_control/gitlab/gitlab_project.py:313]: This change should not be part of this PR, but of a separate PR. (Bugfixes can be backported, new features cannot.)

Comment (josham) [reply plugins/modules/source_control/gitlab/gitlab_project.py]: Here is the separate PR https://github.com/ansible-collections/community.general/pull/3927

It seems like the avatar_path is a recent feature addition and looks unreleased -- so back porting may not be necessary.

File: plugins/doc_fragments/gitlab.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+    # Standard files documentation fragment
+    DOCUMENTATION = r'''
+requirements:
+  - requests (python library https://github.com/psf/requests)
+
+options:
+  api_token:
+    description:
+      - GitLab token for logging in.
```
Comment (felixfontein) [inline plugins/doc_fragments/gitlab.py:19]: I personally like this description better (it is used in a similar amount of modules):
```suggestion
      - GitLab access token with API permissions.
```
What do you think?

File: plugins/doc_fragments/gitlab.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+
+class ModuleDocFragment(object):
+
+    # Standard files documentation fragment
+    DOCUMENTATION = r'''
+requirements:
+  - requests (python library https://github.com/psf/requests)
```
[the remark this answers is not in the document]
Comment (josham) [reply plugins/doc_fragments/gitlab.py:14]: Will change

File: plugins/module_utils/gitlab.py [hunk 7]
```diff
@@ -70,7 +83,16 @@ def gitlab_authentication(module):
             gitlab_instance = gitlab.Gitlab(url=gitlab_url, ssl_verify=validate_certs, email=gitlab_user, password=gitlab_password,
                                             private_token=gitlab_token, api_version=4)
         else:
-            gitlab_instance = gitlab.Gitlab(url=gitlab_url, ssl_verify=validate_certs, private_token=gitlab_token, api_version=4)
+            # We can create an oauth_token using a username and password
+            # https://docs.gitlab.com/ee/api/oauth2.html#authorization-code-flow
+            if gitlab_user:
+                data = {'grant_type': 'password', 'username': gitlab_user, 'password': gitlab_password}
+                resp = requests.post(urljoin(gitlab_url, "oauth/token"), data=data)
```
[the remark this answers is not in the document]
Comment (josham) [reply plugins/module_utils/gitlab.py:90]: Good catch

File: plugins/doc_fragments/gitlab.py (hunk 5 above)
Comment (josham) [reply plugins/doc_fragments/gitlab.py:19]: Will change

File: changelogs/fragments/705-gitlab-auth-support.yml [hunk 8]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - gitlab - add authentication support using oauth_token and job_token (https://github.com/ansible-collections/community.general/issues/705).
```
Comment (felixfontein) [inline changelogs/fragments/705-gitlab-auth-support.yml:2]: I think it's better to mention the names of the new options:
```suggestion
  - gitlab - add more token authentication support with the new options ``api_oauth_token`` and ``api_job_token`` (https://github.com/ansible-collections/community.general/issues/705).
```

**Review by felixfontein — APPROVED** [review]
Besides that, looks good to me!

**Review by russoz — APPROVED** [review]
LGTM

**Review by tadeboro — APPROVED** [review]
Thank you, @josham, for your contribution!

**Comment by felixfontein** [comment]
@josham thanks for implementing this!
@russoz @tadeboro thanks for reviewing!
